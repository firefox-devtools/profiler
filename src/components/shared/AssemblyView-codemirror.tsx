/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
/**
 * This module wraps all the interaction with the CodeMirror API into a
 * AssemblyViewEditor class.
 *
 * This module is intended to be imported asynchronously, so that all the
 * CodeMirror code can be split into a separate bundle chunk.
 *
 * This file implements the following features:
 *  - Display assembly code.
 *  - Display a gutter with:
 *    - "Total" timings for each instruction
 *    - "Self" timings for each instruction
 *    - The address for each instruction
 *  - Highlight assembly code lines which have a non-zero timing, by applying
 *    a cm-nonZeroLine class to them. This highlight line goes across the entire
 *    width of the editor, it covers both the gutter and the main area.
 */
import { EditorView, gutter } from '@codemirror/view';
import { EditorState, StateField, StateEffect } from '@codemirror/state';
import { syntaxHighlighting } from '@codemirror/language';
import { classHighlighter } from '@lezer/highlight';
import clamp from 'clamp';

import {
  AddressTimings,
  Address,
  LineTimings,
  LineNumber,
  NativeSymbolInfo,
  DecodedInstruction,
} from 'firefox-profiler/types';

import { bisectionRight } from 'firefox-profiler/utils/bisect';
import {
  timingsExtension,
  updateTimingsEffect,
  StringMarker,
} from 'firefox-profiler/utils/codemirror-shared';

// An "effect" is like a redux action. This effect is used to replace the value
// of the state field addressToLineMapField.
const updateAddressToLineMapEffect = StateEffect.define<AddressToLineMap>();

// This "state field" stores the current AddressToLineMap. This field allows the
// instructionAddressGutter to map line numbers to addresses.
const addressToLineMapField = StateField.define<AddressToLineMap>({
  create() {
    return new AddressToLineMap([]);
  },
  update(instructionAddresses, transaction) {
    // Get the new value from an effect in the transaction.
    let newSortedAddresses = instructionAddresses;
    for (const effect of transaction.effects) {
      if (effect.is(updateAddressToLineMapEffect)) {
        newSortedAddresses = effect.value;
      }
    }
    return newSortedAddresses;
  },
});

// A gutter which displays the address of each instruction.
const instructionAddressGutter = gutter({
  class: 'cm-instruction-address-gutter',

  // Returns a gutter marker for this line, or null.
  lineMarker(view, line) {
    const lineNumber = view.state.doc.lineAt(line.from).number;
    const map = view.state.field(addressToLineMapField);
    const address = map.lineToAddress(lineNumber);
    return address !== null
      ? new StringMarker(`0x${address.toString(16)}`)
      : null;
  },

  // Returns true if the update affects the instruction addresses in the gutter.
  lineMarkerChange(update) {
    return update.transactions.some((t) =>
      t.effects.some((e) => e.is(updateAddressToLineMapEffect))
    );
  },
});

function instructionsToText(assemblyCode: DecodedInstruction[]): string {
  return assemblyCode.map((instr) => instr.decodedString).join('\n');
}

/**
 * This map is used to convert between instruction addresses and editor line
 * numbers.
 */
class AddressToLineMap {
  // The address of each instruction. This stays constant for the entire lifetime
  // of this AddressToLineMap instance.
  //
  // _instructionAddresses[0] contains the address of the instruction which is
  // displayed in line 1. (Line numbers are 1-based.)
  //
  // The addresses need to be ordered from low to high, so that the binary search
  // works.
  _instructionAddresses: Address[];

  constructor(instructionAddresses: Address[]) {
    this._instructionAddresses = instructionAddresses;
  }

  // Find the line which displays the instruction which covers `address`.
  // `address` doesn't need to be a perfect match for the instruction address;
  // for example, in the example below, address 0x10e4 is mapped to line 3:
  //
  // 1: 0x10da: mov r14, rdi
  // 2: 0x10dd: mov rdi, rsi
  // 3: 0x10e0: call _malloc_usable_size
  // 4: 0x10e5: test rax, rax
  // 5: 0x10e8: je loc_10f6
  addressToLine(address: Address): LineNumber | null {
    const insertionIndex = bisectionRight(this._instructionAddresses, address);
    if (insertionIndex === 0) {
      // address < instructionAddresses[0]
      return null;
    }

    const elementIndex = insertionIndex - 1;
    const lineNumber = elementIndex + 1;
    return lineNumber;
  }

  // Return the address of the instruction which is displayed in line `lineNumber`.
  lineToAddress(lineNumber: LineNumber): Address | null {
    if (lineNumber < 1 || lineNumber > this._instructionAddresses.length) {
      return null;
    }

    const elementIndex = lineNumber - 1;
    return this._instructionAddresses[elementIndex];
  }
}

function getInstructionAddresses(
  assemblyCode: DecodedInstruction[]
): Address[] {
  return assemblyCode.map((instr) => instr.address);
}

// Convert AddressTimings to LineTimings with the help of an AddressToLineMap.
function addressTimingsToLineTimings(
  addressTimings: AddressTimings,
  map: AddressToLineMap
): LineTimings {
  const totalLineHits = new Map();
  for (const [address, hitCount] of addressTimings.totalAddressHits) {
    const line = map.addressToLine(address);
    if (line !== null) {
      const currentHitCount = totalLineHits.get(line) ?? 0;
      totalLineHits.set(line, currentHitCount + hitCount);
    }
  }

  const selfLineHits = new Map();
  for (const [address, hitCount] of addressTimings.selfAddressHits) {
    const line = map.addressToLine(address);
    if (line !== null) {
      const currentHitCount = selfLineHits.get(line) ?? 0;
      selfLineHits.set(line, currentHitCount + hitCount);
    }
  }

  return { totalLineHits, selfLineHits };
}

export class AssemblyViewEditor {
  _view: EditorView;
  _addressToLineMap: AddressToLineMap;
  _addressTimings: AddressTimings;

  // Create a CodeMirror editor and add it as a child element of domParent.
  constructor(
    initialAssemblyCode: DecodedInstruction[],
    nativeSymbol: NativeSymbolInfo,
    addressTimings: AddressTimings,
    domParent: Element
  ) {
    this._addressToLineMap = new AddressToLineMap(
      getInstructionAddresses(initialAssemblyCode)
    );
    this._addressTimings = addressTimings;
    let state = EditorState.create({
      doc: instructionsToText(initialAssemblyCode),
      extensions: [
        timingsExtension,
        addressToLineMapField,
        instructionAddressGutter,
        syntaxHighlighting(classHighlighter),
        EditorState.readOnly.of(true),
        EditorView.editable.of(false),
      ],
    });
    const lineTimings = addressTimingsToLineTimings(
      this._addressTimings,
      this._addressToLineMap
    );
    state = state.update({
      effects: [
        updateAddressToLineMapEffect.of(this._addressToLineMap),
        updateTimingsEffect.of(lineTimings),
      ],
    }).state;
    this._view = new EditorView({
      state,
      parent: domParent,
    });
  }

  setContents(assemblyCode: DecodedInstruction[]) {
    this._addressToLineMap = new AddressToLineMap(
      getInstructionAddresses(assemblyCode)
    );
    const lineTimings = addressTimingsToLineTimings(
      this._addressTimings,
      this._addressToLineMap
    );
    // The CodeMirror way of replacing the entire contents is to insert new text
    // and overwrite the full range of existing text.
    const text = instructionsToText(assemblyCode);
    this._view.dispatch(
      this._view.state.update({
        changes: {
          insert: text,
          from: 0,
          to: this._view.state.doc.length,
        },
      })
    );
    this._view.dispatch({
      effects: [
        updateAddressToLineMapEffect.of(this._addressToLineMap),
        updateTimingsEffect.of(lineTimings),
      ],
    });
  }

  setTimings(addressTimings: AddressTimings) {
    // Update the value of the timings field by dispatching an updateTimingsEffect.
    this._addressTimings = addressTimings;
    const lineTimings = addressTimingsToLineTimings(
      this._addressTimings,
      this._addressToLineMap
    );
    this._view.dispatch({
      effects: updateTimingsEffect.of(lineTimings),
    });
  }

  scrollToLine(lineNumber: number) {
    // Clamp the line number to the document's line count.
    lineNumber = clamp(lineNumber, 1, this._view.state.doc.lines);

    // Convert the line number into a position.
    const pos = this._view.state.doc.line(lineNumber).from;
    // Dispatch the scroll action.
    this._view.dispatch({
      effects: EditorView.scrollIntoView(pos, { y: 'start', yMargin: 0 }),
    });
    // Trigger a measure flush, to work around
    // https://github.com/codemirror/codemirror.next/issues/676
    this._view.coordsAtPos(0);
  }

  scrollToAddress(address: Address) {
    const lineNumber = this._addressToLineMap.addressToLine(address);
    if (lineNumber !== null) {
      this.scrollToLine(lineNumber);
    }
  }

  scrollToAddressWithSpaceOnTop(address: Address, topSpaceLines: number) {
    const lineNumber = this._addressToLineMap.addressToLine(address);
    if (lineNumber !== null) {
      this.scrollToLine(lineNumber - topSpaceLines);
    }
  }
}
