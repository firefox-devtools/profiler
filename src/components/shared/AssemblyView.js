/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
// @flow

import * as React from 'react';

import { ensureExists } from 'firefox-profiler/utils/flow';
import type {
  AddressTimings,
  NativeSymbolInfo,
  DecodedInstruction,
} from 'firefox-profiler/types';
import { mapGetKeyWithMaxValue } from 'firefox-profiler/utils';

import type { AssemblyViewEditor } from './AssemblyView-codemirror';

import './CodeView.css';

const AssemblyViewHeader = () => {
  return (
    <div className="codeViewHeader">
      <span
        className="codeViewHeaderColumn codeViewFixedColumn total"
        title="The “total” sample count includes a summary of every sample where this
line was observed to be on the stack. This includes the time where the
line was actually running, and the time spent in the callers from this
line."
      >
        Total
      </span>
      <span
        className="codeViewHeaderColumn codeViewFixedColumn self"
        title="The “self” sample count only includes the samples where the line was
the end of the stack. If this line called into other functions,
then the “other” functions’ counts are not included. The “self” count is useful
for understanding where time was actually spent in a program."
      >
        Self
      </span>
      <span className="codeViewHeaderColumn codeViewMainColumn assembly"></span>
    </div>
  );
};

type AssemblyViewProps = {
  +timings: AddressTimings,
  +assemblyCode: DecodedInstruction[],
  +disableOverscan: boolean,
  +nativeSymbol: NativeSymbolInfo | null,
  +scrollToHotSpotGeneration: number,
  +hotSpotTimings: AddressTimings,
};

let editorModulePromise: Promise<any> | null = null;

export class AssemblyView extends React.PureComponent<AssemblyViewProps> {
  _ref = React.createRef<HTMLDivElement>();
  _editor: AssemblyViewEditor | null = null;

  /**
   * Scroll to the line with the most hits, based on the timings in
   * timingsForScrolling.
   *
   * How is timingsForScrolling different from this.props.timings?
   * In the current implementation, this.props.timings are always the "global"
   * timings, i.e. they show the line hits for all samples in the current view,
   * regardless of the selected call node. However, when opening the assembly
   * view from a specific call node, you really want to see the code that's
   * relevant to that specific call node, or at least that specific function.
   * So timingsForScrolling are the timings that indicate just the line hits
   * in the selected call node. This means that the "hotspot" will be somewhere
   * in the selected function, and it will even be in the line that's most
   * relevant to that specific call node.
   *
   * Sometimes, timingsForScrolling can be completely empty. This happens, for
   * example, when the assembly view is showing a different file than the
   * selected call node's function's file, for example because we just loaded
   * from a URL and ended up with an arbitrary selected call node.
   * In that case, pick the hotspot from the global line timings.
   */
  _scrollToHotSpot(timingsForScrolling: AddressTimings) {
    const heaviestAddress =
      mapGetKeyWithMaxValue(timingsForScrolling.totalAddressHits) ??
      mapGetKeyWithMaxValue(this.props.timings.totalAddressHits);
    if (heaviestAddress !== undefined) {
      this._scrollToAddressWithSpaceOnTop(heaviestAddress, 5);
    }
  }

  _scrollToAddressWithSpaceOnTop(address: number, topSpaceLines: number) {
    if (this._editor) {
      this._editor.scrollToAddressWithSpaceOnTop(address, topSpaceLines);
    }
  }

  _getAssemblyCodeOrFallback(): DecodedInstruction[] {
    const { assemblyCode } = this.props;
    if (assemblyCode.length !== 0) {
      // We have assembly code for the selected symbol. Good.
      return assemblyCode;
    }

    // We don't have the true assembly code yet, and don't really know
    // at which address each instruction starts.
    // Compute a fallback by getting known addresses from the timings.

    const { timings, nativeSymbol } = this.props;
    const addresses = [...timings.totalAddressHits.keys()];

    // Also include the start address of the symbol, if it's not already present.
    if (
      nativeSymbol !== null &&
      !timings.totalAddressHits.has(nativeSymbol.address)
    ) {
      addresses.push(nativeSymbol.address);
    }

    addresses.sort((a, b) => a - b);

    // Create fallback assembly code where each known address is mapped to an
    // empty string instruction.
    return addresses.map((address) => ({
      address,
      decodedString: '',
    }));
  }

  render() {
    return (
      <div className="assemblyView codeView">
        <AssemblyViewHeader />
        <div className="codeMirrorContainer" ref={this._ref}></div>
      </div>
    );
  }

  componentDidMount() {
    // Load the module with all the @codemirror imports asynchronously, so that
    // it can be split into a separate bundle chunk.
    if (editorModulePromise === null) {
      editorModulePromise = import('./AssemblyView-codemirror');
    }
    (async () => {
      const codeMirrorModulePromise = ensureExists(editorModulePromise);
      const codeMirrorModule = await codeMirrorModulePromise;
      const domParent = this._ref.current;
      if (!domParent) {
        return;
      }
      const { AssemblyViewEditor } = codeMirrorModule;
      const editor = new AssemblyViewEditor(
        this._getAssemblyCodeOrFallback(),
        this.props.nativeSymbol,
        this.props.timings,
        domParent
      );
      this._editor = editor;
      this._scrollToHotSpot(this.props.hotSpotTimings);
    })();
  }

  // CodeMirror's API is not based on React. When our props change, we need to
  // translate those changes into CodeMirror API calls manually.
  componentDidUpdate(prevProps: AssemblyViewProps) {
    if (!this._editor) {
      return;
    }

    let contentsChanged = false;
    if (
      this.props.assemblyCode !== prevProps.assemblyCode ||
      (this.props.assemblyCode.length === 0 &&
        prevProps.assemblyCode.length === 0 &&
        this.props.timings !== prevProps.timings)
    ) {
      this._editor.setContents(this._getAssemblyCodeOrFallback());
      contentsChanged = true;
    }

    if (
      contentsChanged ||
      this.props.scrollToHotSpotGeneration !==
        prevProps.scrollToHotSpotGeneration
    ) {
      this._scrollToHotSpot(this.props.hotSpotTimings);
    }

    if (this.props.timings !== prevProps.timings) {
      this._editor.setTimings(this.props.timings);
    }
  }
}
