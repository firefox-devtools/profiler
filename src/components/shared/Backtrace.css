.backtrace {
  padding: 0;
  margin: 0;
}

.tooltip .backtrace {
  overflow: hidden;
}

.backtraceStackFrame {
  /* Position */
  overflow: hidden;

  /* Box */
  padding: 0;
  margin: 0;
  list-style: none;

  /* Other */
  text-overflow: ellipsis;
  white-space: nowrap;
}

.backtraceStackFrameOrigin {
  margin-left: 10px;
  color: rgb(0 0 0 / 0.55);
  font-style: normal;
}

.backtraceStackFrame_isFrameLabel {
  color: rgb(0 0 0 / 0.6);
}

.backtraceStackFrame_link {
  display: block;
  color: inherit;
  text-decoration: none;
}

.backtraceStackFrame_link:hover {
  background-color: rgb(0 0 0 / 0.05);
}
