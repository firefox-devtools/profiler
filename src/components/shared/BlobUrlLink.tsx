/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

import * as React from 'react';

type Props = {
  // Do not make these props exact, the extra props are passed to the anchor element.
  readonly blob: Blob;
  readonly children: React.ReactNode;
};

type State = {
  url: string;
  prevBlob: Blob | null;
};

/**
 * This component is responsible for converting a Blob into an
 * ObjectUrl. The ObjectUrl strings are not GCed, so this component
 * does the proper thing of cleaning up after itself as the component
 * is mounted, updated, and unmounted.
 */
export class BlobUrlLink extends React.PureComponent<Props, State> {
  override state: State = {
    url: '',
    prevBlob: null,
  };

  static getDerivedStateFromProps(props: Props, state: State) {
    if (props.blob === state.prevBlob) {
      return null;
    }
    if (state.prevBlob) {
      URL.revokeObjectURL(state.url);
    }
    return {
      url: URL.createObjectURL(props.blob),
      prevBlob: props.blob,
    };
  }

  override componentWillUnmount() {
    URL.revokeObjectURL(this.state.url);
  }

  override render() {
    const {
      blob,
      children,
      ...rest
    } = this.props;

    // This component must be an <a> rather than a <button> as the download attribute
    // allows users to download the profile.
    return (
      <a href={this.state.url} {...rest}>
        {children}
      </a>
    );
  }
}
