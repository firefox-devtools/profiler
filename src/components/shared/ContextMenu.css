/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

.react-contextmenu {
  z-index: 4; /* needs to be on a higher level than .overflowEdgeIndicatorEdge */
  display: none;
  min-width: 160px;
  max-width: 600px;
  padding: 5px 0;
  border-radius: 3px;
  margin: 2px 0 0;
  background-color: #fff;
  box-shadow:
    0 0 0 0.5px rgb(0 0 0 / 0.1),
    0 10px 12px rgb(0 0 0 / 0.3);
  color: #000;
  font-size: 12px;
  text-align: left;
  user-select: none;
}

.react-contextmenu-separator {
  border-bottom: 1px solid #ddd;
  margin: 6px 0;
}

/* max-height for context menus in the panel changed to 90vh for better zoom experience  */
.react-contextmenu.react-contextmenu--visible {
  display: block;
  max-height: 90vh;
}

.react-contextmenu-item:not(.react-contextmenu-submenu) {
  display: flex;
  overflow: hidden;
  align-items: center;
  padding: 2px 12px;
  cursor: default;
  line-height: 1.5;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.react-contextmenu-item.react-contextmenu-item--active,
.react-contextmenu-item--selected {
  border-color: var(--blue-60);
  background-color: var(--blue-60);
  color: #fff;
  text-decoration: none;
}

.react-contextmenu-item.react-contextmenu-item--disabled,
.react-contextmenu-item.react-contextmenu-item--disabled:hover {
  background-color: transparent;
  color: #888;
}

.react-contextmenu-item--divider {
  padding: 2px 0;
  border-bottom: 1px solid rgb(0 0 0 / 0.15);
  margin-bottom: 3px;
}

/* Submenu arrow */
.react-contextmenu-item.react-contextmenu-submenu
  > .react-contextmenu-item::after {
  position: absolute;
  top: 5px;
  right: 7px;
  width: 0;
  height: 0;
  border: 6px solid transparent;
  border-left-color: #000;
  content: '';
}

/* Submenu arrow highlighted color */
.react-contextmenu-item:hover.react-contextmenu-submenu
  > .react-contextmenu-item::after,
.react-contextmenu-item.react-contextmenu-submenu
  > .react-contextmenu-item--active::after {
  border-left-color: highlighttext;
}

.react-contextmenu-item.checkable {
  position: relative;
  padding-right: 21px;
}

.react-contextmenu-item.indented {
  padding-left: 35px;
}

.react-contextmenu-item.checked:not(.react-contextmenu-item--disabled)::before {
  position: absolute;
  top: 6px;
  right: 8px;
  display: block;
  width: 3px;
  height: 6px;
  border: solid #000;
  border-width: 0 2px 2px 0;
  content: '';
  transform: rotate(45deg);
}

.react-contextmenu-item.react-contextmenu-item--active.checked,
.react-contextmenu-item--selected.checked {
  border-color: #fff;
}

.react-contextmenu-item--disabled.react-contextmenu-item--selected {
  background-color: var(--blue-50-a30);
}

.react-contextmenu-item--selected.checked:not(
    .react-contextmenu-item--disabled
  )::before {
  /* Invert the colors of the checkmark when selected */
  border-color: #fff;
}

/* Use this wrapper for a content that needs to take the available space or if
 * it contains other HTML elements (eg <strong>) */
.react-contextmenu-item-content {
  overflow: hidden;
  min-width: 0;
  flex: 1;
  text-overflow: ellipsis;
}

/* Use a span with this class to add an icon to an item. */
.react-contextmenu-icon {
  width: 16px;
  height: 16px;
  flex: none;
  background: center no-repeat;
  margin-inline-end: 8px;
  pointer-events: auto;
}

/* Soften a bit the style of highlighted words */
.react-contextmenu-item > strong {
  color: #38383d;
}

.react-contextmenu-item--selected > strong {
  color: #fff;
}

@media (forced-colors: active) {
  .react-contextmenu {
    border: 1px solid CanvasText;
  }

  .react-contextmenu-item:not(.react-contextmenu-item--disabled):hover {
    background-color: SelectedItem;
    color: SelectedItemText;
  }

  .react-contextmenu-item.react-contextmenu-item--disabled,
  .react-contextmenu-item.react-contextmenu-item--disabled:hover {
    color: GrayText;
  }
}
