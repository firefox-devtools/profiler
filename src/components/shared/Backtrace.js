/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

import React from 'react';
import classNames from 'classnames';
import { getBacktraceItemsForStack } from 'firefox-profiler/profile-logic/transforms';

import type {
  CategoryList,
  Thread,
  IndexIntoStackTable,
  ImplementationFilter,
} from 'firefox-profiler/types';

import './Backtrace.css';

type Props = {|
  +thread: Thread,
  // Tooltips will want to only show a certain number of stacks, while the sidebars
  // can show all of the stacks.
  +maxStacks: number,
  +stackIndex: IndexIntoStackTable,
  +implementationFilter: ImplementationFilter,
  +categories: CategoryList,
|};

export function Backtrace(props: Props) {
  const { stackIndex, thread, implementationFilter, maxStacks, categories } =
    props;
  const funcNamesAndOrigins = getBacktraceItemsForStack(
    stackIndex,
    implementationFilter,
    thread
  );

  if (funcNamesAndOrigins.length) {
    return (
      <ul className="backtrace">
        {funcNamesAndOrigins
          // Truncate the stacks
          .slice(0, maxStacks)
          .map(({ funcName, origin, isFrameLabel, category }, i) => (
            <li
              key={i}
              className={classNames('backtraceStackFrame', {
                backtraceStackFrame_isFrameLabel: isFrameLabel,
              })}
            >
              <span
                className={`colored-border category-color-${categories[category].color}`}
                title={categories[category].name}
              />
              {funcName}
              <em className="backtraceStackFrameOrigin">{origin}</em>
            </li>
          ))}
        {funcNamesAndOrigins.length > maxStacks
          ? [
              <span
                key={funcNamesAndOrigins.length}
                className="colored-border ellipsis"
              />,
              'â€¦',
            ]
          : null}
      </ul>
    );
  }
  return (
    <div className="backtrace">
      (The stack is empty because all of its frames are filtered out by the
      implementation filter. Switch the implementation filter in the call tree
      to see more frames.)
    </div>
  );
}
