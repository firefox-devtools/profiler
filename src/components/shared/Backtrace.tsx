/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

import classNames from 'classnames';
import { Localized } from '@fluent/react';
import { getBacktraceItemsForStack } from 'firefox-profiler/profile-logic/transforms';
import { getFunctionName } from 'firefox-profiler/profile-logic/function-info';

import type {
  CategoryList,
  Thread,
  IndexIntoStackTable,
  ImplementationFilter,
} from 'firefox-profiler/types';

import './Backtrace.css';

type Props = {
  readonly thread: Thread;
  // Tooltips will want to only show a certain number of stacks, while the sidebars
  // can show all of the stacks.
  readonly maxStacks: number;
  readonly stackIndex: IndexIntoStackTable;
  readonly implementationFilter: ImplementationFilter;
  readonly categories: CategoryList;
};

export function Backtrace(props: Props) {
  const { stackIndex, thread, implementationFilter, maxStacks, categories } =
    props;
  const funcNamesAndOrigins = getBacktraceItemsForStack(
    stackIndex,
    implementationFilter,
    thread
  );

  if (funcNamesAndOrigins.length) {
    return (
      <ul className="backtrace">
        {funcNamesAndOrigins
          // Truncate the stacks
          .slice(0, maxStacks)
          .map(
            ({ funcName, origin, isFrameLabel, category, inlineDepth }, i) => (
              <li
                key={i}
                className={classNames('backtraceStackFrame', {
                  backtraceStackFrame_isFrameLabel: isFrameLabel,
                })}
              >
                <span
                  className={`colored-border category-color-${categories[category].color}`}
                  title={categories[category].name}
                />
                {inlineDepth > 0 ? (
                  <Localized
                    id="Backtrace--inlining-badge"
                    vars={{ function: getFunctionName(funcName) }}
                    attrs={{ title: true }}
                  >
                    <span className="backtraceBadge inlined" title="inlined">
                      (inlined)
                    </span>
                  </Localized>
                ) : null}
                {funcName}
                <em className="backtraceStackFrameOrigin">{origin}</em>
              </li>
            )
          )}
        {funcNamesAndOrigins.length > maxStacks
          ? [
              <span
                key={funcNamesAndOrigins.length}
                className="colored-border ellipsis"
              />,
              'â€¦',
            ]
          : null}
      </ul>
    );
  }
  return (
    <div className="backtrace">
      (The stack is empty because all of its frames are filtered out by the
      implementation filter. Switch the implementation filter in the call tree
      to see more frames.)
    </div>
  );
}
