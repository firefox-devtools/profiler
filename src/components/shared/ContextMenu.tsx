/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

import { PureComponent } from 'react';
import type { ComponentProps } from 'react';
import { ContextMenu as ReactContextMenu } from '@firefox-devtools/react-contextmenu';

import './ContextMenu.css';

type Props = ComponentProps<typeof ReactContextMenu>;

export class ContextMenu extends PureComponent<Props> {
  _mouseDownHandler = (event: React.MouseEvent<HTMLDivElement>): void => {
    // This prevents from stealing the focus from where it was.
    event.preventDefault();
  };

  override render() {
    return (
      <div onMouseDown={this._mouseDownHandler}>
        <ReactContextMenu {...this.props}>
          {this.props.children ? this.props.children : <div />}
        </ReactContextMenu>
      </div>
    );
  }
}
