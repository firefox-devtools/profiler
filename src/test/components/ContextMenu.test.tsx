/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
import { render } from 'firefox-profiler/test/fixtures/testing-library';
import { ContextMenu } from '../../components/shared/ContextMenu';

describe('ContextMenu', function () {
  it('correctly renders the context-menu with the props that were passed through', () => {
    const { container } = render(
      <ContextMenu
        id="contextMenu"
        className="context-menu"
        data={{}}
        onShow={() => null}
      />
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
