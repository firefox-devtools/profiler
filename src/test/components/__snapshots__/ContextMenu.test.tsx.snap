// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`ContextMenu correctly renders the context-menu with the props that were passed through 1`] = `
<div>
  <nav
    class="react-contextmenu context-menu"
    role="menu"
    style="position: fixed; opacity: 0; pointer-events: none;"
    tabindex="-1"
  >
    <div />
  </nav>
</div>
`;
